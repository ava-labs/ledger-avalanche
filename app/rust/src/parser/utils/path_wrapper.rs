/*******************************************************************************
*   (c) 2018-2024 Zondax AG
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
********************************************************************************/
use bolos::crypto::bip32::BIP32Path;
use core::{mem::MaybeUninit, ptr::addr_of_mut};

use crate::{
    constants::BIP32_PATH_SUFFIX_DEPTH,
    parser::{FromBytes, ObjectList, ParserError, U32_SIZE},
};
use nom::{bytes::complete::take, number::complete::be_u8};

/// A simple wrapper around BIP32Path objects, in order to use it along
/// with ObjectList.
pub struct PathWrapper<const PATH_DEPTH: usize>(BIP32Path<PATH_DEPTH>);

impl<'b, const PATH_DEPTH: usize> FromBytes<'b> for PathWrapper<PATH_DEPTH> {
    #[inline(never)]
    fn from_bytes_into(
        input: &'b [u8],
        out: &mut MaybeUninit<Self>,
    ) -> Result<&'b [u8], nom::Err<ParserError>> {
        crate::sys::zemu_log_stack("PathWrapper::from_bytes_into\x00");

        let (_, num_components) = be_u8(input)?;
        // compute the amount of bytes for this path
        let bytes = (num_components as usize * U32_SIZE) + 1;

        let (rem, raw_path) = take(bytes)(input)?;

        // get the path and also validates
        let path = BIP32Path::read(raw_path).map_err(|_| ParserError::InvalidPath)?;

        //good ptr and no uninit reads
        let out = out.as_mut_ptr();
        unsafe {
            addr_of_mut!((*out).0).write(path);
        }

        Ok(rem)
    }
}

impl<const PATH_DEPTH: usize> PathWrapper<PATH_DEPTH> {
    pub fn path(&self) -> BIP32Path<PATH_DEPTH> {
        self.0
    }
}

pub fn parse_path_list<'a>(
    path_list_uninit: &mut MaybeUninit<ObjectList<'a, PathWrapper<BIP32_PATH_SUFFIX_DEPTH>>>,
    data: &'a [u8],
) -> Result<&'a [u8], ParserError> {
    // Parse the number of paths
    let (rem, num_paths) = be_u8(data)?;

    // Try to initialize the ObjectList into the provided MaybeUninit
    let rem = ObjectList::new_into_with_len(rem, path_list_uninit, num_paths as usize)
        .map_err(|_| ParserError::InvalidPath)?;

    // Safely assume initialization
    let _path_list = unsafe { path_list_uninit.assume_init() };

    // Return the remainder of the input data
    Ok(rem)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::constants::MAX_BIP32_PATH_DEPTH;
    use crate::parser::ObjectList;

    const NUM_PATHS: usize = 3;
    const LIST: &[u8] = &[
        // "0/0", "0/1", "1/100"
        0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x01, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x64,
    ];

    #[test]
    fn test_path_list() {
        let mut list: MaybeUninit<ObjectList<PathWrapper<MAX_BIP32_PATH_DEPTH>>> =
            MaybeUninit::uninit();
        let _ = ObjectList::new_into_with_len(LIST, &mut list, NUM_PATHS).unwrap();
        let list = unsafe { list.assume_init() };
        list.iter()
            .for_each(|path| assert_eq!(path.path().components().len(), 2));
        assert_eq!(list.iter().count(), NUM_PATHS);
    }
}
