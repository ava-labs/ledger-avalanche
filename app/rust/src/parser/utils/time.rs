/*******************************************************************************
*   (c) 2018-2024 Zondax AG
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
********************************************************************************/
use crate::parser::{u64_to_str, u8_to_str, FORMATTED_STR_DATE_LEN, U64_FORMATTED_SIZE};
use crate::sys::{pic_str, PIC};
use arrayvec::ArrayVec;
use arrayvec::CapacityError;

const MONTH_DAYS: &[u8; 12] = &[31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];

const YEAR_LOOKUP: &[u32; 400] = &[
    0, 365, 730, 1096, 1461, 1826, 2191, 2557, 2922, 3287, 3652, 4018, 4383, 4748, 5113, 5479,
    5844, 6209, 6574, 6940, 7305, 7670, 8035, 8401, 8766, 9131, 9496, 9862, 10227, 10592, 10957,
    11323, 11688, 12053, 12418, 12784, 13149, 13514, 13879, 14245, 14610, 14975, 15340, 15706,
    16071, 16436, 16801, 17167, 17532, 17897, 18262, 18628, 18993, 19358, 19723, 20089, 20454,
    20819, 21184, 21550, 21915, 22280, 22645, 23011, 23376, 23741, 24106, 24472, 24837, 25202,
    25567, 25933, 26298, 26663, 27028, 27394, 27759, 28124, 28489, 28855, 29220, 29585, 29950,
    30316, 30681, 31046, 31411, 31777, 32142, 32507, 32872, 33238, 33603, 33968, 34333, 34699,
    35064, 35429, 35794, 36160, 36525, 36890, 37255, 37621, 37986, 38351, 38716, 39082, 39447,
    39812, 40177, 40543, 40908, 41273, 41638, 42004, 42369, 42734, 43099, 43465, 43830, 44195,
    44560, 44926, 45291, 45656, 46021, 46387, 46752, 47117, 47482, 47847, 48212, 48577, 48942,
    49308, 49673, 50038, 50403, 50769, 51134, 51499, 51864, 52230, 52595, 52960, 53325, 53691,
    54056, 54421, 54786, 55152, 55517, 55882, 56247, 56613, 56978, 57343, 57708, 58074, 58439,
    58804, 59169, 59535, 59900, 60265, 60630, 60996, 61361, 61726, 62091, 62457, 62822, 63187,
    63552, 63918, 64283, 64648, 65013, 65379, 65744, 66109, 66474, 66840, 67205, 67570, 67935,
    68301, 68666, 69031, 69396, 69762, 70127, 70492, 70857, 71223, 71588, 71953, 72318, 72684,
    73049, 73414, 73779, 74145, 74510, 74875, 75240, 75606, 75971, 76336, 76701, 77067, 77432,
    77797, 78162, 78528, 78893, 79258, 79623, 79989, 80354, 80719, 81084, 81450, 81815, 82180,
    82545, 82911, 83276, 83641, 84006, 84371, 84736, 85101, 85466, 85832, 86197, 86562, 86927,
    87293, 87658, 88023, 88388, 88754, 89119, 89484, 89849, 90215, 90580, 90945, 91310, 91676,
    92041, 92406, 92771, 93137, 93502, 93867, 94232, 94598, 94963, 95328, 95693, 96059, 96424,
    96789, 97154, 97520, 97885, 98250, 98615, 98981, 99346, 99711, 100076, 100442, 100807, 101172,
    101537, 101903, 102268, 102633, 102998, 103364, 103729, 104094, 104459, 104825, 105190, 105555,
    105920, 106286, 106651, 107016, 107381, 107747, 108112, 108477, 108842, 109208, 109573, 109938,
    110303, 110669, 111034, 111399, 111764, 112130, 112495, 112860, 113225, 113591, 113956, 114321,
    114686, 115052, 115417, 115782, 116147, 116513, 116878, 117243, 117608, 117974, 118339, 118704,
    119069, 119435, 119800, 120165, 120530, 120895, 121260, 121625, 121990, 122356, 122721, 123086,
    123451, 123817, 124182, 124547, 124912, 125278, 125643, 126008, 126373, 126739, 127104, 127469,
    127834, 128200, 128565, 128930, 129295, 129661, 130026, 130391, 130756, 131122, 131487, 131852,
    132217, 132583, 132948, 133313, 133678, 134044, 134409, 134774, 135139, 135505, 135870, 136235,
    136600, 136966, 137331, 137696, 138061, 138427, 138792, 139157, 139522, 139888, 140253, 140618,
    140983, 141349, 141714, 142079, 142444, 142810, 143175, 143540, 143905, 144271, 144636, 145001,
    145366, 145732,
];

#[cfg_attr(any(test, feature = "derive-debug"), derive(Debug))]
pub enum TimeError {
    InvalidTimestamp,
    BufferTooSmall,
}

impl<T> From<CapacityError<T>> for TimeError {
    fn from(_: CapacityError<T>) -> Self {
        TimeError::BufferTooSmall
    }
}

// this is used to add a leading 0
// to a single digit number, to "emulate"
// format!("{:0>2}", 1); which returns 01
// this is used to format month, days, hours, minutes
// and seconds when formatting timestamps.
macro_rules! add_padding {
    ($num:expr, $string:expr) => {
        if $num < 10 {
            $string.try_push(b'0')?;
        }
    };
}

#[cfg_attr(any(test, feature = "derive-debug"), derive(Debug))]
pub struct Date {
    day: u8,
    month: u8,
    year: u32,
    hour: u8,
    min: u8,
    sec: u8,
}

/// Conversts a unix `timestamp`
/// returns a date
pub fn timestamp_to_date(timestamp: i64) -> Result<Date, TimeError> {
    let month_days = PIC::new(MONTH_DAYS).into_inner();
    let year_lookup = PIC::new(YEAR_LOOKUP).into_inner();

    let mut t = timestamp;

    let tm_sec = (t % 60) as u8;
    t -= tm_sec as i64;
    t /= 60;

    let tm_min = (t % 60) as u8;
    t -= tm_min as i64;
    t /= 60;

    let tm_hour = (t % 24) as u8;
    t -= tm_hour as i64;
    t /= 24;

    // Look up tm_year
    let mut tm_year = 0;

    while tm_year < year_lookup.len() && year_lookup[tm_year] <= t as _ {
        tm_year += 1;
    }

    if tm_year == 0 || tm_year == year_lookup.len() {
        return Err(TimeError::InvalidTimestamp);
    }
    tm_year -= 1;

    let mut tm_day: u16 = (t as u32 - year_lookup[tm_year] + 1) as u16;
    tm_year += 1970;

    // Get day/month
    let leap = ((tm_year % 4 == 0) && (tm_year % 100 != 0 || tm_year % 400 == 0)) as u8;

    let mut tm_mon = 0;
    for i in 0..12 {
        tm_mon = i;
        let mut tmp = month_days[tm_mon];
        tmp += if tm_mon == 1 { leap } else { 0 };

        if tm_day <= tmp as _ {
            break;
        }
        tm_day -= tmp as u16;
    }
    tm_mon += 1;

    Ok(Date {
        day: tm_day as u8,
        month: tm_mon as _,
        year: tm_year as _,
        hour: tm_hour,
        min: tm_min,
        sec: tm_sec,
    })
}

/// Conversts a unix `timestamp`
/// returns a formatted date string  on success
pub fn timestamp_to_str_date(
    timestamp: i64,
) -> Result<ArrayVec<u8, FORMATTED_STR_DATE_LEN>, TimeError> {
    let date = timestamp_to_date(timestamp)?;

    let mut date_str = ArrayVec::<_, FORMATTED_STR_DATE_LEN>::new();
    let mut num_buff = [0; U64_FORMATTED_SIZE + 2];

    // separators
    let dash = b'-';
    let space = b' ';
    let colon = b':';
    let utc = pic_str!(b"UTC"!);

    // year, does not require adding leading
    // zeroes
    let num =
        u64_to_str(date.year as _, &mut num_buff[..]).map_err(|_| TimeError::BufferTooSmall)?;
    date_str.try_extend_from_slice(num)?;
    date_str.try_push(dash)?;
    // month
    add_padding!(date.month, date_str);
    let num =
        u8_to_str(date.month as _, &mut num_buff[..]).map_err(|_| TimeError::BufferTooSmall)?;
    date_str.try_extend_from_slice(num)?;
    date_str.try_push(dash)?;
    // day
    add_padding!(date.day, date_str);
    let num = u8_to_str(date.day, &mut num_buff[..]).map_err(|_| TimeError::BufferTooSmall)?;
    date_str.try_extend_from_slice(num)?;
    date_str.try_push(space)?;

    // time
    // hour
    add_padding!(date.hour, date_str);
    let num = u8_to_str(date.hour, &mut num_buff[..]).map_err(|_| TimeError::BufferTooSmall)?;
    date_str.try_extend_from_slice(num)?;
    date_str.try_push(colon)?;
    // min
    add_padding!(date.min, date_str);
    let num = u8_to_str(date.min, &mut num_buff[..]).map_err(|_| TimeError::BufferTooSmall)?;
    date_str.try_extend_from_slice(num)?;
    date_str.try_push(colon)?;
    // seconds
    add_padding!(date.sec, date_str);
    let num = u8_to_str(date.sec, &mut num_buff[..]).map_err(|_| TimeError::BufferTooSmall)?;
    date_str.try_extend_from_slice(num)?;
    date_str.try_push(space)?;

    // it is redundant to have Utc appended at the end
    // as by definition unix-timestamp is Utc, but this
    // keeps compatibility with legacy app
    date_str.try_extend_from_slice(&utc[..])?;

    Ok(date_str)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn timestamp_to_str() {
        use time::format_description;
        use time::OffsetDateTime;

        let timestamp = [1657089945, 55471502, 1663087639];

        for t in timestamp {
            let date = OffsetDateTime::from_unix_timestamp(t).unwrap();
            let format =
                format_description::parse("[year]-[month]-[day] [hour]:[minute]:[second] UTC")
                    .unwrap();

            let date_str = date.format(&format).unwrap();

            let test_date = timestamp_to_str_date(t).unwrap();
            let test_date = std::str::from_utf8(test_date.as_slice()).unwrap();

            assert_eq!(&date_str, test_date);
        }
    }
}
