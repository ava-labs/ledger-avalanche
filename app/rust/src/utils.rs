/*******************************************************************************
*   (c) 2018-2024 Zondax AG
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
********************************************************************************/
#![allow(dead_code, unused_macros)]

use bolos::PIC;

mod apdu_unwrap;
pub use apdu_unwrap::*;

mod apdu_wrapper;
pub use apdu_wrapper::*;

pub mod convert_to_rs;
pub use convert_to_rs::convert_der_to_rs;

mod buffer_upload;
pub use buffer_upload::*;

mod app_mode;
pub use app_mode::*;

pub mod blind_sign_toggle;
mod ui;

#[cfg(test)]
#[macro_export]
macro_rules! assert_error_code {
    ($tx:expr, $buffer:ident, $expected:expr) => {
        let pos: usize = $tx as _;
        let actual: ApduError = (&$buffer[pos - 2..pos]).try_into().unwrap();
        assert_eq!(actual, $expected);
    };
}

/// This function returns the index of the first null byte in the slice
#[cfg(test)]
pub fn strlen(s: &[u8]) -> usize {
    let mut count = 0;
    while let Some(&c) = s.get(count) {
        if c == 0 {
            return count;
        }
        count += 1;
    }

    panic!("byte slice did not terminate with null byte, s: {:x?}", s)
}

/// This function returns the index of the
/// first null byte in the slice or the total len of the slice,
/// whichever comes first
pub fn rs_strlen(s: &[u8]) -> usize {
    let mut count = 0;
    while let Some(&c) = s.get(count) {
        if c == 0 {
            return count;
        }
        count += 1;
    }

    s.len()
}

pub struct OutputBufferTooSmall;
pub fn hex_encode(
    input: impl AsRef<[u8]>,
    output: &mut [u8],
) -> Result<usize, OutputBufferTooSmall> {
    let input = input.as_ref();

    if input.len() * 2 > output.len() {
        return Err(OutputBufferTooSmall);
    }

    const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";

    let table = PIC::new(HEX_CHARS_LOWER).into_inner();
    for (byte, out) in input.iter().zip(output.chunks_mut(2)) {
        let high = table[((byte & 0xf0) >> 4) as usize];
        let low = table[(byte & 0xf) as usize];

        //number of items guaranteed
        // as we checked the size beforehand so
        // output will always be at least the right length
        // to encode input
        out[0] = high;
        out[1] = low;
    }

    Ok(input.len() * 2)
}

pub fn bs58_encode(
    input: impl AsRef<[u8]>,
    output: &mut [u8],
) -> Result<usize, OutputBufferTooSmall> {
    const ALPHABET_ENCODE: &[u8; 58] =
        b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
    let table = PIC::new(ALPHABET_ENCODE).into_inner();

    let input = input.as_ref();
    let mut index = 0;

    for &val in input.iter() {
        let mut carry = val as usize;
        for byte in &mut output[..index] {
            carry += (*byte as usize) << 8;
            *byte = (carry % 58) as u8;
            carry /= 58;
        }
        while carry > 0 {
            if index == output.len() {
                return Err(OutputBufferTooSmall);
            }
            output[index] = (carry % 58) as u8;
            index += 1;
            carry /= 58;
        }
    }

    for _ in input.iter().take_while(|v| **v == 0) {
        if index == output.len() {
            return Err(OutputBufferTooSmall);
        }
        output[index] = 0;
        index += 1;
    }

    for val in &mut output[..index] {
        *val = table[*val as usize];
    }

    output.get_mut(..index).apdu_unwrap().reverse();
    Ok(index)
}

/// Reads a byte slice preprended with the slice len
pub fn read_slice(input: &[u8]) -> Option<(usize, &[u8])> {
    let len = input.first()?;
    let len = *len as usize;

    input.get(1..1 + len).map(|bytes| (1 + len, bytes))
}

#[cfg(test)]
mod maybe_null_terminated_to_string {
    use core::str::Utf8Error;
    use std::borrow::ToOwned;
    use std::ffi::{CStr, CString};
    use std::string::String;

    ///This trait is a utility trait to convert a slice of bytes into a CString
    ///
    /// If the string is nul terminated already then no null termination is added
    pub trait MaybeNullTerminatedToString {
        fn to_string_with_check_null(&self) -> Result<String, Utf8Error>;
    }

    impl MaybeNullTerminatedToString for &[u8] {
        fn to_string_with_check_null(&self) -> Result<String, Utf8Error> {
            //attempt to make a cstr first
            if let Ok(cstr) = CStr::from_bytes_with_nul(self) {
                return cstr.to_owned().into_string().map_err(|e| e.utf8_error());
            }

            //in the case above,
            // we could be erroring due to a null byte in the middle
            // or a null byte _missing_ at the end
            //
            //but here we'll error for a null byte at the end or a null byte in the middle
            match CString::new(self.to_vec()) {
                Ok(cstring) => cstring.into_string().map_err(|e| e.utf8_error()),
                Err(err) => {
                    // so with the above error, we can only be erroring here only with a null byte in the middle
                    let nul_pos = err.nul_position();
                    //truncate the string
                    CStr::from_bytes_with_nul(&self[..=nul_pos])
                        //we can't be erroring for a missing null byte at the end,
                        // and also can't error due to a null byte in the middle,
                        // because this is literally the smaller substring to be terminated
                        .unwrap()
                        .to_owned()
                        .into_string()
                        .map_err(|e| e.utf8_error())
                }
            }
        }
    }

    impl<const S: usize> MaybeNullTerminatedToString for [u8; S] {
        fn to_string_with_check_null(&self) -> Result<String, Utf8Error> {
            (&self[..]).to_string_with_check_null()
        }
    }
}

#[macro_export]
/// Convert the return of Show::show into something more usable for apdu handlers
///
/// sets `tx` to the amount returned if given,
/// otherwise tx is returned only on success and discarded on failure
macro_rules! show_ui {
    ($show:expr, $tx:ident) => {{
        let result = $show;
        match result {
            Ok((size, err)) if err == $crate::constants::ApduError::Success as u16 => {
                let size = size as _;
                *$tx = size;
                Ok(())
            }
            Ok((size, err)) => {
                use core::convert::TryInto;
                let size = size as _;
                *$tx = size;

                match err.try_into() {
                    Ok(err) => Err(err),
                    Err(_) => Err($crate::constants::ApduError::ExecutionError),
                }
            }
            Err(_) => Err($crate::constants::ApduError::ExecutionError),
        }
    }};
    ($show:expr) => {{
        let result = $show;
        match result {
            Ok((size, err)) if err == $crate::constants::ApduError::Success as u16 => Ok(size as _),
            Ok((_, err)) => {
                use core::convert::TryInto;

                match err.try_into() {
                    Ok(err) => Err(err),
                    Err(_) => Err($crate::constants::ApduError::ExecutionError),
                }
            }
            Err(_) => Err($crate::constants::ApduError::ExecutionError),
        }
    }};
}
