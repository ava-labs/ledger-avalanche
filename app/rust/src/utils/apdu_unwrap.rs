/*******************************************************************************
*   (c) 2018-2024 Zondax AG
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
********************************************************************************/

pub trait ApduPanic: Sized {
    type Item;

    fn apdu_unwrap(self) -> Self::Item;

    fn apdu_expect(self, s: &str) -> Self::Item;
}

impl<T, E> ApduPanic for Result<T, E> {
    type Item = T;

    #[inline]
    fn apdu_unwrap(self) -> Self::Item {
        match self {
            Ok(t) => t,
            Err(_) => panic!(),
        }
    }

    #[inline]
    fn apdu_expect(self, _: &str) -> Self::Item {
        match self {
            Ok(t) => t,
            Err(_) => panic!(),
        }
    }
}

impl<T> ApduPanic for Option<T> {
    type Item = T;

    #[inline]
    fn apdu_unwrap(self) -> Self::Item {
        match self {
            Some(t) => t,
            None => panic!(),
        }
    }

    #[inline]
    fn apdu_expect(self, _: &str) -> Self::Item {
        match self {
            Some(t) => t,
            None => panic!(),
        }
    }
}
