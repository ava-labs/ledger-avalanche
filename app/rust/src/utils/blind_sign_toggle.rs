/*******************************************************************************
*   (c) 2018-2024 Zondax AG
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
********************************************************************************/
use cfg_if::cfg_if;

#[cfg(feature = "blind-sign-toggle")]
mod impls {
    #[repr(C)]
    pub struct BlindSignToggle {
        pub toggle: bool,
        pub message: [i8; 9],
    }

    cfg_if::cfg_if! {
        if #[cfg(any(unix, windows))] {
            /// Provide a mock for tests
            #[allow(non_upper_case_globals)]
            pub static mut blind_sign: BlindSignToggle = BlindSignToggle {
                toggle: true,
                message: [0; 9],
            };
        } else {
            extern "C" {
                ///Link to the C code
                pub static mut blind_sign: BlindSignToggle;
            }
        }
    }
}

/// Returns if blind signing is enabled in this execution
pub fn blind_sign_enabled() -> bool {
    cfg_if! {
        if #[cfg(feature = "blind-sign-toggle")] {
            //safe: guaranteed no data races
            unsafe { bolos::PIC::new(&impls::blind_sign).into_inner().toggle }
        } else {
            false
        }
    }
}
