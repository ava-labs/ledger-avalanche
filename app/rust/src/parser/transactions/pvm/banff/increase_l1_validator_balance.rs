/*******************************************************************************
*   (c) 2023 Zondax AG
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
********************************************************************************/
use crate::utils::hex_encode;
use crate::{
    handlers::handle_ui_message,
    parser::{
        nano_avax_to_fp_str, BaseTxFields, DisplayableItem, FromBytes, Header, ParserError,
        PvmOutput, PVM_INCREASE_L1_VALIDATOR_BALANCE, U64_FORMATTED_SIZE
    },
};
use bolos::PIC;
use core::{mem::MaybeUninit, ptr::addr_of_mut};
use nom::bytes::complete::{tag, take};
use nom::number::complete::be_u64;
use zemu_sys::ViewError;

pub const VALIDATION_ID_LEN: usize = 32;

#[derive(Clone, Copy, PartialEq, Eq)]
#[repr(C)]
#[cfg_attr(test, derive(Debug))]
pub struct IncreaseL1ValidatorBalanceTx<'b> {
    pub tx_header: Header<'b>,
    pub base_tx: BaseTxFields<'b, PvmOutput<'b>>,
    pub validation_id: &'b [u8; VALIDATION_ID_LEN],
    pub balance: u64,
}

// Info at https://github.com/ava-labs/avalanchejs/blob/master/src/utils/getBurnedAmountByTx.ts
impl<'b> IncreaseL1ValidatorBalanceTx<'b> {
    fn fee(&'b self) -> Result<u64, ParserError> {
        let sum_inputs = self.base_tx.sum_inputs_amount()?;

        let base_outputs = self.base_tx.sum_outputs_amount()?;

        let mut fee = sum_inputs
            .checked_sub(base_outputs)
            .ok_or(ParserError::OperationOverflows)?;

        fee = fee
            .checked_sub(self.balance)
            .ok_or(ParserError::OperationOverflows)?;
        Ok(fee)
    }
}

impl<'b> FromBytes<'b> for IncreaseL1ValidatorBalanceTx<'b> {
    fn from_bytes_into(
        input: &'b [u8],
        out: &mut MaybeUninit<Self>,
    ) -> Result<&'b [u8], nom::Err<ParserError>> {
        crate::sys::zemu_log_stack("IncreaseL1ValidatorBalanceTx::from_bytes_into\x00");

        let (rem, _) = tag(PVM_INCREASE_L1_VALIDATOR_BALANCE.to_be_bytes())(input)?;

        let out = out.as_mut_ptr();

        // tx header
        let tx_header = unsafe { &mut *addr_of_mut!((*out).tx_header).cast() };
        let rem = Header::from_bytes_into(rem, tx_header)?;

        // base_tx
        let base_tx = unsafe { &mut *addr_of_mut!((*out).base_tx).cast() };
        let rem = BaseTxFields::<PvmOutput>::from_bytes_into(rem, base_tx)?;

        // validation_id
        let (rem, validation_id) = take(32usize)(rem)?; // Read 32 bytes
        let validation_id = arrayref::array_ref!(validation_id, 0, 32);

        // balance
        let (rem, balance) = be_u64(rem)?;

        unsafe {
            addr_of_mut!((*out).validation_id).write(validation_id);
            addr_of_mut!((*out).balance).write(balance);
        }

        Ok(rem)
    }
}

impl<'b> DisplayableItem for IncreaseL1ValidatorBalanceTx<'b> {
    fn num_items(&self) -> Result<u8, ViewError> {
        // tx_info, validation_id, amount, fee
        Ok(4u8)
    }

    fn render_item(
        &self,
        item_n: u8,
        title: &mut [u8],
        message: &mut [u8],
        page: u8,
    ) -> Result<u8, zemu_sys::ViewError> {
        use bolos::pic_str;

        let mut buffer = [0; U64_FORMATTED_SIZE + 2];

        match item_n {
            0 => {
                let label = pic_str!(b"IncreaseL1ValBal");
                title[..label.len()].copy_from_slice(label);
                let content = pic_str!(b"Transaction");
                handle_ui_message(content, message, page)
            }
            1 => {
                let prefix = pic_str!(b"0x"!);
                let label = pic_str!(b"Validator");
                title[..label.len()].copy_from_slice(label);

                // prefix
                let mut out = [0; VALIDATION_ID_LEN * 2 + 2];
                let mut sz = prefix.len();
                out[..prefix.len()].copy_from_slice(&prefix[..]);

                sz += hex_encode(self.validation_id, &mut out[prefix.len()..])
                    .map_err(|_| ViewError::Unknown)?;

                handle_ui_message(&out[..sz], message, page)
            }
            2 => {
                let label = pic_str!(b"Amount(AVAX)");
                title[..label.len()].copy_from_slice(label);

                let balance_buff = nano_avax_to_fp_str(self.balance, &mut buffer[..])
                    .map_err(|_| ViewError::Unknown)?;

                handle_ui_message(balance_buff, message, page)
            }
            3 => {
                let label = pic_str!(b"Fee(AVAX)");
                title[..label.len()].copy_from_slice(label);

                let fee = self.fee().map_err(|_| ViewError::Unknown)?;
                let fee_buff =
                    nano_avax_to_fp_str(fee, &mut buffer[..]).map_err(|_| ViewError::Unknown)?;

                handle_ui_message(fee_buff, message, page)
            }
            _ => Err(ViewError::NoData),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::prelude::v1::*;

    use crate::parser::snapshots_common::ReducedPage;
    use zuit::Page;

    const DATA: &[u8] = &[];

    include!("testvectors/increase_l1_validator_balance.rs");
    #[test]
    fn parse_increase_l1_validator_balance() {
        let validation_id = &[
            0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef, 0x01, 0x23, 0x45, 0x67, 0x89, 0xab,
            0xcd, 0xef, 0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef, 0x01, 0x23, 0x45, 0x67,
            0x89, 0xab, 0xcd, 0xef,
        ];

        let (_, tx) =
            IncreaseL1ValidatorBalanceTx::from_bytes(INCREASE_L1_VALIDATOR_BALANCE_DATA).unwrap();
        assert_eq!(tx.validation_id, validation_id);
        assert_eq!(tx.balance, 1000000000);
    }

    #[test]
    #[cfg_attr(miri, ignore)]
    fn ui_increase_l1_validator_balance() {
        println!("-------------------- Increase L1 Validator Balance TX ------------------------");
        let (_, tx) =
            IncreaseL1ValidatorBalanceTx::from_bytes(INCREASE_L1_VALIDATOR_BALANCE_DATA).unwrap();

        let items = tx.num_items().expect("Overflow?");

        let mut pages = Vec::<Page<18, 1024>>::with_capacity(items as usize);
        for i in 0..items {
            let mut page = Page::default();

            tx.render_item(i as _, &mut page.title, &mut page.message, 0)
                .unwrap();

            pages.push(page);
        }

        let reduced = pages.iter().map(ReducedPage::from).collect::<Vec<_>>();
        insta::assert_debug_snapshot!(reduced);
    }
}
