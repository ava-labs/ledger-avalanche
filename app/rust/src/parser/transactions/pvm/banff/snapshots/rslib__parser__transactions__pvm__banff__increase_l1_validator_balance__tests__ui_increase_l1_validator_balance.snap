---
source: src/parser/transactions/pvm/banff/increase_l1_validator_balance.rs
assertion_line: 204
expression: reduced
---
[
    "IncreaseL1ValBal": "Transaction",
    "Validator": "0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
    "Amount(AVAX)": "1",
    "Fee(AVAX)": "1152921505.606846974",
]
